File: Medium/22. Generate Parentheses/generateParentheses.java
public class GenerateParentheses {

    public List<String> generateParenthesis(int n) {
        // Resultant list
        List<String> result = new ArrayList<>();
        /// Recursively generate parentheses
        generateParenthesis(result, "", 0, 0, n);
        return result;
    }

    private void generateParenthesis(List<String> result, String s, int open, int close, int n) {
        // Base case
        if (open == n && close == n) {
            result.add(s);
            return;
        }
        // If the number of open parentheses is less than the given n
        if (open < n) {
            generateParenthesis(result, s + "(", open + 1, close, n);
        }
        // If we need more close parentheses to balance
        if (close < open) {
            generateParenthesis(result, s + ")", open, close + 1, n);
        }
    }
}

File: Medium/22. Generate Parentheses/generateParentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */

// using recurrsion to solve the problems

var generateParenthesis = function(n) {
    
    let stack = [];
    recurrsion(0, 0, '', stack, n);
    return stack;
};

function recurrsion(openCount, closeCount, str, stack, n) {
    
    // If open parenthese + close parenthese = 2n, then finish
    if(openCount === n && closeCount === n) {
        stack.push(str);
        return;
    }
    
    // add open parenthese if open parenthese < n
    if(openCount < n)
       recurrsion(openCount+1,  closeCount, str+'(', stack, n);
    
    // add close parenthese if close parenthese < open parenthese
    if(closeCount < openCount) 
       recurrsion(openCount,  closeCount+1, str+')', stack, n);
    
}
